src/runtime/client-event-manager.js

```diff
@@ -15,33 +15,133 @@
  */
 
 import {FilterResult} from '../api/client-event-manager-api';
+import {AnalyticsEvent,EventOriginator} from '../proto/api_messages';
+import {isObject, isFunction, isEnumValue, isBoolean} from '../utils/types';
 
-/** @implements {../api/client-event-manager-api.ClientEventManagerApi}*/
+/**
+ * Helper function to describe an issue with an event object
+ * @param {!string} valueName
+ * @param {?*} value
+ * @returns {!string}
+ */
+function createEventErrorMessage(valueName, value) {
+  return 'Event has an invalid ' + valueName + '(' + value + ')';
+}
+
+/**
+ * Throws an error if the event is invalid.
+ * @param {!../api/client-event-manager-api.ClientEvent} event
+ */
+function validateEvent(event) {
+  if (!isObject(event)) {
+    throw new Error('Event must be a valid object');
+  }
+
+  if (!isEnumValue(AnalyticsEvent, event.eventType)) {
+    throw new Error(createEventErrorMessage('eventType', event.eventType));
+  }
+
+  if (!isEnumValue(EventOriginator, event.eventOriginator)) {
+    throw new Error(createEventErrorMessage('eventOriginator',
+        event.eventOriginator));
+  }
+
+  if (!isObject(event.additionalParameters)
+      && event.additionalParameters !== null) {
+    if (event.additionalParameters !== undefined) {
+      throw new Error(createEventErrorMessage('additionalParameters',
+          event.additionalParameters));
+    }
+    event.additionalParameters = null;
+  }
+
+
+  if (event.isFromUserAction !== null && !isBoolean(event.isFromUserAction)) {
+    throw new Error(createEventErrorMessage('isFromUserAction',
+        event.isFromUserAction));
+  }
+}
+
+/** @implements {../api/client-event-manager-api.ClientEventManagerApi} */
 export class ClientEventManager {
   constructor() {
+    /** @private {!Array<function(!../api/client-event-manager-api.ClientEvent)>} */
+    this.listeners_ = [];
+
+    /** @private {!Array<function(!../api/client-event-manager-api.ClientEvent):!FilterResult>} */
+    this.filterers_ = [];
+
+    /** @private {?Promise} */
+    this.lastAction_ = null;
   }
 
   /**
    * @overrides
    */
-  registerEventListener(unusedCallback) {
+  registerEventListener(listener) {
+    if (!isFunction(listener)) {
+      throw new Error('Event manager listeners must be a function');
+    }
+    this.listeners_.push(listener);
   }
 
   /**
    * @overrides
    */
-  registerEventFilterer(unusedCallback) {
+  registerEventFilterer(filterer) {
+    if (!isFunction(filterer)) {
+      throw new Error('Event manager filterers must be a function');
+    }
+    this.filterers_.push(filterer);
   }
 
   /**
    * @overrides
    */
   logEvent(event) {
```

Comment:
This is still as asynchronous we'd like it to be. The correct implementation is as what you had it before.

this.lastAction_ =  validateEvent(event).then(() => {
 // filter event
 // invoke callbacks
});

Let's revisit why you want lastAction to track the filtering. Once again, as a principle, tests should not dictate the implementation.

Reply:
We discussed my modified approach.